components/datetime/src/format/time_zone.rs

```diff
@@ -37,12 +37,82 @@ where
     T: TimeZoneInput,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write_pattern(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
+        write_zone(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
     }
 }
 
-// TODO(#622) Make this public once TimeZoneFormat is public.
-pub(crate) fn write_pattern<T, W>(
+pub(crate) fn write_zone<T, W>(
+    time_zone_format: &TimeZoneFormat,
+    time_zone: &T,
+    w: &mut W,
+) -> Result<(), Error>
+where
+    T: TimeZoneInput,
+    W: fmt::Write + ?Sized,
+{
+    match (&time_zone_format.patterns, &time_zone_format.config) {
+        (Some(_), None) => write_pattern(time_zone_format, time_zone, w),
+        (None, Some(_)) => write_config(time_zone_format, time_zone, w),
+        (Some(_), Some(_)) => {
+            unreachable!("TimeZoneFormat should have a configuration or a pattern, but found both")
+        }
+        (None, None) => unreachable!(
+            "TimeZoneFormat should have either a configuration or a pattern, but found neither"
+        ),
+    }
+}
+
+fn write_config<T, W>(
+    time_zone_format: &TimeZoneFormat,
+    time_zone: &T,
+    w: &mut W,
+) -> Result<(), Error>
```

Comment:
Suggestion: To avoid the need for the `.expect()` a few lines down, you could add an argument `config: &TimeZoneFormatConfig` which you populate at the call site where you have the variable.